### PR TITLE
allow `type` and `type alias`

### PR DESCRIPTION
**Pull request by stoeffel**
allows to use sum- and product-types in examples.

```elm
{-|
    type Animal
        = Dog
        | Cat

    double Dog
    --> (Dog, Dog)
-}
double : a -> (a, a)
double a =
    (a, a)
```

related https://github.com/stoeffel/elm-verify/pull/9 (cc @michaeljones)

### CODE REVIEW

**Comment by stoeffel** [comment]
@eeue56 yes, totally. I was planning to create a file for each example. I think that will lead to less confusing errors. What do you think?

**Comment by eeue56** [comment]
Maybe if the files are only made if there's a type involved? Otherwise it's increasing the IO considerably (and Elm will compile things slower)

**Comment by stoeffel** [comment]
### TODO
- [ ] pass all files to write to js once

File: src/VerifyExamples.elm [hunk 1]
```diff
@@ -60,11 +60,13 @@ update msg model =
...
             let
-                generatedTestFileText =
+                generatedTests =
                     Parser.parse fileText
-                        |> Compiler.compile moduleName
+                        |> List.concatMap (Compiler.compile moduleName)
             in
-                model ! [ writeFile ( moduleName, generatedTestFileText ) ]
+            ( model
+            , writeFiles generatedTests
```
Comment (stoeffel) [inline src/VerifyExamples.elm]: @eeue56 changed it to send everything to js and write these generatedTests in serial

File: run-tests.sh [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #! /bin/bash -ex
-TEST_COUNT=21
+TEST_COUNT=18
```
Comment (stoeffel) [inline run-tests.sh]: Yes! it's finally green again

**Comment by stoeffel** [comment]
@eeue56 would love another review if you have time. I think I've addressed all your comments from the other PR and I've cleaned some more things. I still don't really like `Compile.elm`, but I think I'll clean that up in a new PR.

**Comment by stoeffel** [comment]
reminder : please squash this PR. commits are useless (😿 )

File: src/VerifyExamples/Ast.elm [hunk 3]
```diff
@@ -75,18 +47,38 @@ isExpectiation x =
             False
 
 
-isLocalFunction : Ast -> Bool
-isLocalFunction x =
-    case x of
-        LocalFunction _ _ ->
-            True
+group : List Ast -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
```
Comment (stoeffel) [inline src/VerifyExamples/Ast.elm]: maybe `groupExtras`

File: src/VerifyExamples/Ast.elm [hunk 4]
```diff
@@ -97,40 +89,39 @@ astToString ast =
...
-        Expression _ str ->
-            str
 
-        Func _ str ->
-            str
+groupToAst : List IntermediateAst -> Maybe Ast
+groupToAst xs =
+    case xs of
+        (IAst.MaybeExpression x) :: rest ->
+            Assertion (String.join "\n" <| x :: List.map IAst.toString rest)
```
Comment (stoeffel) [inline src/VerifyExamples/Ast.elm]: use String.Util.unlines here

File: src/VerifyExamples/IntermediateAst.elm [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+expectationRegex =
+    Regex.regex "^\\s{4}\\-\\->\\s(.*)"
+
+
+localFunctionRegex : Regex
+localFunctionRegex =
+    Regex.regex "^\\s{4}(\\w+\\s:\\s.*)"
+
+
+oneOf : List (a -> Maybe b) -> a -> Maybe b
```
Comment (stoeffel) [inline src/VerifyExamples/IntermediateAst.elm]: create a module

File: .travis.yml [hunk 6]
```diff
@@ -32,7 +32,7 @@ install:
   - npm install
 
 script:
-  - npm test
+  - $TRAVIS_BUILD_DIR/sysconfcpus/bin/sysconfcpus -n 2 npm test
```
Comment (eeue56) [inline .travis.yml]: using `-n 1` is actually more performant than -n 2 :D

File: src/VerifyExamples.elm (hunk 1 above)
Comment (eeue56) [reply src/VerifyExamples.elm]: 👍

File: src/VerifyExamples/Ast.elm [hunk 7]
```diff
@@ -75,18 +48,38 @@ isExpectiation x =
...
+group ast =
+    groupHelp ast { localFunctions = [], imports = [], types = [] }
 
-        _ ->
-            False
+
+groupHelp :
+    List Ast
+    -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
+    -> { localFunctions : List Ast, imports : List Ast, types : List Ast }
```
Comment (eeue56) [inline src/VerifyExamples/Ast.elm]: 🎨 why not give a name to this type?

File: src/VerifyExamples/Ast.elm [hunk 8]
```diff
@@ -97,40 +90,49 @@ astToString ast =
...
-        Func _ str ->
-            str
+groupToAst : List IntermediateAst -> Maybe Ast
+groupToAst xs =
+    case xs of
+        (IAst.MaybeExpression x) :: rest ->
+            (x :: List.map IAst.toString rest)
+                |> unlines
+                |> Assertion
+                |> Just
```
Comment (eeue56) [inline src/VerifyExamples/Ast.elm]: This common pattern could be a function

File: src/VerifyExamples/Comment.elm [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        (Just comment) :: (Just functionName) :: _ ->
+            Just { comment = comment, functionName = functionName }
+
+        _ ->
+            Nothing
+
+
+commentRegex : Regex
+commentRegex =
+    Regex.regex "({-[^]*?-})\x0D?\n([^\x0D\n(\\s:)(\\s=))]*)\\s[:=]"
```
Comment (eeue56) [inline src/VerifyExamples/Comment.elm]: 😱 this is very hard to read

Comment (stoeffel) [reply src/VerifyExamples/Comment.elm]: yes, it's horrible. I will do something....

File: src/VerifyExamples/Ast.elm (hunk 8 above)
Comment (stoeffel) [reply src/VerifyExamples/Ast.elm]: :+1:

File: .travis.yml (hunk 6 above)
Comment (stoeffel) [reply .travis.yml]: oh, yes. forgot about that

File: src/VerifyExamples/Ast.elm (hunk 7 above)
Comment (stoeffel) [reply src/VerifyExamples/Ast.elm]: sure

File: src/VerifyExamples/Comment.elm [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+        _ ->
+            Nothing
+
+
+commentRegex : Regex
+commentRegex =
+    Regex.regex <|
+        String.join ""
+            [ "({-[^]*?-})" -- anything between comments
```
Comment (stoeffel) [inline src/VerifyExamples/Comment.elm]: @eeue56 does this help?

Comment (zwilias) [reply src/VerifyExamples/Comment.elm]: `String.concat` is a useful little `String.join ""` alternative - less thinking about what's going on here.

Comment (stoeffel) [reply src/VerifyExamples/Comment.elm]: d'oh! will fix.

**Comment by stoeffel** [comment]
@eeue56 addressed all comments.

File: src/String/Util.elm [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+unlines : List String -> String
+unlines =
+    String.join "\n"
+
+
+escape : String -> String
+escape =
+    Regex.replace All (regex "\\\\") (\_ -> "\\\\")
+        >> Regex.replace All (regex "\\\"") (\_ -> "\\\"")
+        >> Regex.replace All (regex "\\s\\s+") (\_ -> " ")
```
Comment (eeue56) [inline src/String/Util.elm]: ⚠️ potential runtime errors here

File: src/VerifyExamples/Comment.elm [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+commentRegex : Regex
+commentRegex =
+    Regex.regex <|
+        String.concat
+            [ "({-[^]*?-})" -- anything between comments
+            , newline
+            , "([^\\s(" ++ newline ++ ")]*)" -- anything that is not a space or newline
+            , "\\s[:=]" -- until ` :` or ` =`
+            ]
```
Comment (eeue56) [inline src/VerifyExamples/Comment.elm]: 👍

File: src/String/Util.elm (hunk 11 above)
Comment (stoeffel) [reply src/String/Util.elm]: will extract. good catch!

**Pull request merged**